[WC-962] Implement Heatmap chart

Keraito opened the pull request:
## Checklist
- Contains unit tests ✅ ❌
- Contains breaking changes ✅ ❌
- Contains Atlas changes ✅ ❌
- Compatible with: MX 7️⃣, 8️⃣, 9️⃣

#### Web specific
- Contains e2e tests ❌
- Is accessible ✅ 
- Compatible with Studio ✅ 
- Cross-browser compatible ✅ 

#### Feature specific
- Comply with designs ✅ 
- Comply with PM's requirements ✅ 

## This PR contains
- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
Add heatmap chart

Keraito committed:
feat(heatmap-chart-web): setup folder structure

Keraito committed:
feat(heatmap-chart-web): implement heatmap features

Keraito committed:
feat(piw-utils-internal): abstract attribute value compare function

Keraito committed:
feat(pie-chart-web): use piw-utils-internal compareFn

Keraito committed:
feat(heatmap-chart-web): implement data sorting

Keraito committed:
feat(shared-charts): export Annotations type

Keraito committed:
feat(heatmap-chart-web): implement scale and annotations related properties

Keraito committed:
feat(heatmap-chart-web): add tests

Keraito committed:
feat(heatmap-chart-web): skip e2e tests

Keraito committed:
feat(heatmap-chart-web): move common properties to own tab and make valuesColor conditional

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/src/HeatMap.editorConfig.ts:
```diff
@@ -0,0 +1,47 @@
+    return null;
+}
+
+export function check(_values: HeatMapPreviewProps): Problem[] {
```
If no checks are being not maybe we can just remove the whole function

Keraito replied:
Fine with me, I always assume I'll need it 😄

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/src/HeatMap.tsx:
```diff
@@ -0,0 +1,112 @@
+        () => ({
+            ...heatmapChartLayoutOptions,
+            annotations: props.showValues
+                ? heatmapChartData[0].z.reduce<HeatMapAnnotation[]>(
```
What is data is empty?

Keraito replied:
Based on [the return value](https://github.com/mendix/widgets-resources/pull/1234/files#diff-8f234248ed7a972a957d8549363c5a66a08fe89618e0c05de7617c5b413569a7R138-R154) of the `useHeatMapDataSeries` hook, that's not possible. There will always be an array with 1 item

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/src/HeatMap.xml:
```diff
@@ -0,0 +1,244 @@
+                <property key="verticalSortAttribute" type="attribute" required="false" dataSource="seriesDataSource">
+                    <caption>Vertical sort attribute</caption>
+                    <description>
+                        Attribute to use for sorting the data. Sorting can only be used when data source is ‘Database’.
```
Maybe `Attribute to be used for sorting the data` sounds better.

diego-antonelli replied:
If thats the case then lets close this discussion.

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/src/__tests__/HeatMap.spec.tsx:
```diff
@@ -0,0 +1,237 @@
+        );
+    }
+
+    it("visualizes data as a heatmap chart", () => {
```
Can we add a test case when there is no data?

Keraito replied:
sure

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/src/hooks/data.ts:
```diff
@@ -0,0 +1,188 @@
+              .sort((colour1, colour2) => colour1.valuePercentage - colour2.valuePercentage)
+              .map(colors => [Math.abs(colors.valuePercentage / 100), colors.colour])
+        : [
+              [0, "#17347B"],
```
Are these default colors being provided by Ana?

Keraito replied:
No, copied from last chart

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/src/package.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<package xmlns="http://www.mendix.com/package/1.0/">
+    <clientModule name="HeatMap" version="1.0.0" xmlns="http://www.mendix.com/clientModule/1.0/">
```
I think it should be v3.0.0

Keraito replied:
👍

diego-antonelli commented on packages/pluggableWidgets/heatmap-chart-web/package.json:
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "heatmap-chart-web",
+  "widgetName": "HeatMap",
+  "version": "1.0.0",
```
Version is incorrect as HeatMap was already published in the past. It should be 3.0.0
https://github.com/mendix/charts/blob/master/src/package.xml

diego-antonelli commented on packages/tools/piw-utils-internal/src/utils/AttributeValueUtils.ts:
```diff
@@ -0,0 +1,36 @@
+import Big from "big.js";
+import { compareAsc } from "date-fns";
```
We need to make sure `date-fns` is available in all Charts package as the output of `piw-utils-internal` is just a transpiled js file.

Keraito replied:
oh no.. will take a look into this

diego-antonelli reviewed (changes requested):
Very good PR.

I've noticed Dark icons are not defined here. Better check with Ana.

Keraito committed:
feat(heatmap-chart-web): address review comments

- bump version to 3.0.0 to match existing upload
- add test to verify that empty data doesn't error
- remove unused editorConfig check function

Keraito committed:
feat(): add comment to attributevalue util about date-fns dependency

Keraito commented on packages/pluggableWidgets/heatmap-chart-web/package.json:
```diff
@@ -43,6 +43,7 @@
     "@mendix/shared-charts": "^1.0.0",
-    "classnames": "^2.2.6"
+    "classnames": "^2.2.6",
+    "date-fns": "^2.24.0"
```
Didn't lock this yet cuz it will cause hoisting conflicts with date-grid. I think it's better if we fix it in a separate PR

Keraito committed:
feat(heatmap-chart-web): remove unused import

Keraito committed:
feat(heatmap-chart-web): fix indentations of XML

Keraito committed:
feat(heatmap-chart-web): remove irrelevant seriesName attribute

Keraito committed:
feat(heatmap-chart-web): default annotation text to empty string

Keraito committed:
feat(heatmap-chart-web): explicitly state that 0% and 100% need to be defined for scale colors

Keraito committed:
feat(heatmap-chart-web): increase distance between y axis title and vlaues

Keraito committed:
feat(heatmap-chart-web): force plotly to accept matrix for hovertext

Keraito committed:
feat(heatmap-chart-web): remove leftover typing from copied widget

Keraito commented:
@diego-antonelli had to rebase, therefore the force push. These are the new commits based on feedback by @leonardomendix:

![image](https://user-images.githubusercontent.com/5955441/151148452-fd4ec72c-1928-4551-8c42-62d0230a6fb0.png)

Keraito committed:
feat(shared-charts): fix tests
